File: new_backend/express-backend/src/routes/onboarding.ts
import express, { Request, Response, NextFunction, Router } from 'express';
import { PrismaClient } from '@prisma/client';
import multer from 'multer';
import axios from 'axios';
import { provisionAgentNumber, deprovisionAgentNumber } from '../services/twilioProvisioningService';

const router: Router = express.Router();
const upload = multer({ dest: 'uploads/' });

// Extend Request interface
declare global {
  namespace Express {
    interface Request {
      tenantId: string;
      userId: string;
    }
  }
}

// ─── Company search endpoint ──────────────────────────────────────────────────
// Proxies to Clearbit Autocomplete (free, no auth) to search any real company.
// Falls back to empty array if Clearbit is unreachable.
router.get('/company-search', async (req: Request, res: Response) => {
  const q = ((req.query.q as string) || '').trim();
  if (!q || q.length < 1) {
    return res.json({ companies: [] });
  }
  try {
    const response = await axios.get(
      `https://autocomplete.clearbit.com/v1/companies/suggest?query=${encodeURIComponent(q)}`,
      { timeout: 5000 }
    );
    const companies = (response.data || []).slice(0, 10).map((item: any) => ({
      id:          item.domain || item.name.toLowerCase().replace(/\s+/g, '-'),
      name:        item.name,
      domain:      item.domain || '',
      industry:    '',        // Clearbit autocomplete doesn't expose industry
      description: item.domain || '',
    }));
    res.json({ companies });
  } catch (err) {
    console.error('Clearbit search failed:', err);
    res.json({ companies: [] });
  }
});

// ─── Company profile setup ────────────────────────────────────────────────────
// GET: return existing company profile from tenant settings
router.get('/company', async (req: Request, res: Response) => {
  try {
    const prisma: PrismaClient = req.app.get('prisma');
    const tenant = await prisma.tenant.findUnique({ where: { id: req.tenantId } });
    const settings = (tenant?.settings as any) || {};
    res.json({
      company_name: settings.companyName || null,
      industry: settings.industry || null,
      use_case: settings.useCase || null,
      website_url: settings.websiteUrl || null,
      description: settings.description || null,
    });
  } catch (error) {
    console.error('Error fetching company profile:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// POST: saves company metadata to tenant.settings AND triggers a background scrape
// of the company website so that content lands in the vector store immediately.
router.post('/company', async (req: Request, res: Response) => {
  try {
    const prisma: PrismaClient = req.app.get('prisma');
    const { company_name, industry, use_case, website_url, description } = req.body;

    // Persist to Postgres tenant record
    await prisma.tenant.update({
      where: { id: req.tenantId },
      data: {
        settings: {
          companyName: company_name,
          industry: industry,
          useCase: use_case,
          websiteUrl: website_url || null,
          description: description || null,
        },
      },
    });

    // If a website URL was provided, kick off a background ingestion job so
    // company knowledge is available right away for RAG.
    let scrapeJobId: string | null = null;
    if (website_url) {
      try {
        const ingestionUrl = process.env.FASTAPI_URL || 'http://localhost:8001';
        const scrapeRes = await axios.post(`${ingestionUrl}/ingest/company`, {
          tenantId: req.tenantId,
          website_url,
          company_name: company_name || '',
          company_description: description || undefined,
          industry: industry || undefined,
          use_case: use_case || undefined,
        });
        scrapeJobId = scrapeRes.data?.job_id || null;
        console.log(`Company scrape job started: ${scrapeJobId} for tenant ${req.tenantId}`);
      } catch (scrapeErr) {
        // Non-fatal: log but don't block the onboarding response
        console.error('Failed to trigger company scrape:', scrapeErr);
      }
    }

    res.json({ success: true, scrapeJobId });
  } catch (error) {
    console.error('Error saving company profile:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// ─── Proxy: ingestion job status ─────────────────────────────────────────────
// Frontend polls this to show scraping progress without talking directly to FastAPI.
router.get('/scrape-status/:jobId', async (req: Request, res: Response) => {
  try {
    const ingestionUrl = process.env.FASTAPI_URL || 'http://localhost:8001';
    const response = await axios.get(`${ingestionUrl}/status/${req.params.jobId}`);
    res.json(response.data);
  } catch (error) {
    console.error('Error fetching scrape status:', error);
    res.status(500).json({ error: 'Failed to fetch scrape status' });
  }
});

// ─── Proxy: company knowledge list ───────────────────────────────────────────
// Returns chunks from ChromaDB that belong to the company_profile source.
router.get('/company-knowledge', async (req: Request, res: Response) => {
  try {
    const ingestionUrl = process.env.FASTAPI_URL || 'http://localhost:8001';
    const response = await axios.get(
      `${ingestionUrl}/knowledge/company/${req.tenantId}`
    );
    res.json(response.data);
  } catch (error) {
    console.error('Error fetching company knowledge:', error);
    res.status(500).json({ error: 'Failed to fetch company knowledge' });
  }
});

// ─── Proxy: delete a knowledge chunk ─────────────────────────────────────────
router.delete('/company-knowledge/:chunkId', async (req: Request, res: Response) => {
  try {
    const ingestionUrl = process.env.FASTAPI_URL || 'http://localhost:8001';
    await axios.delete(
      `${ingestionUrl}/knowledge/${req.tenantId}/${req.params.chunkId}`
    );
    res.json({ deleted: true });
  } catch (error) {
    console.error('Error deleting knowledge chunk:', error);
    res.status(500).json({ error: 'Failed to delete chunk' });
  }
});

// Agent creation
router.post('/agent', async (req: Request, res: Response) => {
  try {
    const prisma: PrismaClient = req.app.get('prisma');
    const { name, role, templateId, description, channels, brandId } = req.body;

    // If brandId provided, verify it belongs to the same tenant
    if (brandId) {
      const brand = await prisma.brand.findFirst({ where: { id: brandId, tenantId: req.tenantId } });
      if (!brand) {
        return res.status(400).json({ error: 'Brand not found or does not belong to this tenant' });
      }
    }

    // Create the agent with optional template and brand reference
    const agent = await prisma.agent.create({
      data: {
        name,
        description: description || role || undefined,
        templateId: templateId || undefined,
        brandId: brandId || undefined,
        channels: channels || undefined,
        userId: req.userId,
        tenantId: req.tenantId,
      },
    });

    // Create the corresponding agent configuration (non-fatal if schema differs)
    try {
      const tenantSettings = (
        await prisma.tenant.findUnique({ where: { id: req.tenantId } })
      )?.settings as Record<string, any> | null;

      await prisma.agentConfiguration.create({
        data: {
          agentId: agent.id,
          templateId: templateId || undefined,
          agentName: name,
          agentRole: role || undefined,
          agentDescription: description || undefined,
          communicationChannels: channels || undefined,
          companyName: tenantSettings?.companyName || undefined,
          industry: tenantSettings?.industry || undefined,
          primaryUseCase: tenantSettings?.useCase || undefined,
        },
      });
    } catch (configErr) {
      console.warn('Failed to create agent configuration (non-fatal):', configErr);
    }

    res.json({ agent_id: agent.id });
  } catch (error) {
    console.error('Error creating agent:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Knowledge upload
router.post('/knowledge', upload.array('files'), async (req: Request, res: Response) => {
  try {
    const files = req.files as Express.Multer.File[];
    const { websites, faq_text } = req.body;

    // Forward to ingestion service
    const ingestionUrl = process.env.FASTAPI_URL || 'http://localhost:8001';

    const formData = new FormData();
    files?.forEach((file) => {
      // Convert multer file to blob for FormData
      const fs = require('fs');
      const blob = new Blob([fs.readFileSync(file.path)], { type: file.mimetype });
      formData.append('files', blob, file.originalname);
    });

    if (websites) {
      formData.append('websites', JSON.stringify(websites));
    }

    if (faq_text) {
      formData.append('faq_text', faq_text);
    }

    formData.append('tenant_id', req.tenantId);
    formData.append('user_id', req.userId);

    const response = await axios.post(`${ingestionUrl}/ingest`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });

    res.json({ success: true, jobId: response.data.job_id });
  } catch (error) {
    console.error('Error uploading knowledge:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Voice configuration — save voiceId to agent configuration
router.post('/voice', async (req: Request, res: Response) => {
  try {
    const prisma: PrismaClient = req.app.get('prisma');
    const { voice, tone, language, personality } = req.body;

    // Find the tenant's most recent agent and update its configuration
    const agent = await prisma.agent.findFirst({
      where: { tenantId: req.tenantId },
      orderBy: { createdAt: 'desc' },
    });

    if (agent) {
      await prisma.agentConfiguration.upsert({
        where: { agentId: agent.id },
        update: {
          voiceId: voice || undefined,
          responseTone: tone || undefined,
          preferredResponseStyle: personality || undefined,
        },
        create: {
          agentId: agent.id,
          voiceId: voice || undefined,
          responseTone: tone || undefined,
          preferredResponseStyle: personality || undefined,
        },
      });
    }

    res.json({ success: true });
  } catch (error) {
    console.error('Error configuring voice:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Channel setup
router.post('/channels', async (req: Request, res: Response) => {
  try {
    // For now, just return success
    res.json({ success: true });
  } catch (error) {
    console.error('Error setting up channels:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Agent configuration
router.post('/agent-config', async (req: Request, res: Response) => {
  try {
    const prisma: PrismaClient = req.app.get('prisma');
    const {
      agent_name, agent_role, agent_description,
      personality_traits, communication_channels,
      preferred_response_style, response_tone, voice_id,
      company_name, industry, primary_use_case,
    } = req.body;

    // Find the tenant's most recent agent
    const agent = await prisma.agent.findFirst({
      where: { tenantId: req.tenantId },
      orderBy: { createdAt: 'desc' },
    });

    if (agent) {
      await prisma.agentConfiguration.upsert({
        where: { agentId: agent.id },
        update: {
          agentName: agent_name || undefined,
          agentRole: agent_role || undefined,
          agentDescription: agent_description || undefined,
          personalityTraits: personality_traits || undefined,
          communicationChannels: communication_channels || undefined,
          preferredResponseStyle: preferred_response_style || undefined,
          responseTone: response_tone || undefined,
          voiceId: voice_id || undefined,
          companyName: company_name || undefined,
          industry: industry || undefined,
          primaryUseCase: primary_use_case || undefined,
        },
        create: {
          agentId: agent.id,
          agentName: agent_name || undefined,
          agentRole: agent_role || undefined,
          agentDescription: agent_description || undefined,
          personalityTraits: personality_traits || undefined,
          communicationChannels: communication_channels || undefined,
          preferredResponseStyle: preferred_response_style || undefined,
          responseTone: response_tone || undefined,
          voiceId: voice_id || undefined,
          companyName: company_name || undefined,
          industry: industry || undefined,
          primaryUseCase: primary_use_case || undefined,
        },
      });
    }

    res.json({
      success: true,
      message: 'Agent configured successfully',
      agent_id: agent?.id || 'unknown',
      chroma_collection: `collection_${req.tenantId}`,
    });
  } catch (error) {
    console.error('Error configuring agent:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Deploy agent — provision a real Twilio phone number
router.post('/deploy', async (req: Request, res: Response) => {
  try {
    const prisma: PrismaClient = req.app.get('prisma');
    const { agent_id, country_code } = req.body;

    if (!agent_id) {
      return res.status(400).json({ error: 'agent_id is required' });
    }

    // Verify the agent belongs to this tenant
    const agent = await prisma.agent.findFirst({
      where: { id: agent_id, tenantId: req.tenantId },
    });
    if (!agent) {
      return res.status(404).json({ error: 'Agent not found' });
    }

    // If the agent already has a number, return it
    if (agent.phoneNumber) {
      return res.json({ success: true, phone_number: agent.phoneNumber, already_provisioned: true });
    }

    // Pre-flight: ensure tenant has Twilio credentials configured
    const tenant = await prisma.tenant.findUnique({ where: { id: req.tenantId } });
    const tenantSettings = (tenant?.settings as Record<string, any>) || {};
    if (!tenantSettings.twilioAccountSid || !tenantSettings.twilioAuthToken) {
      return res.status(400).json({
        error: 'Twilio credentials not configured. Please add your Twilio Account SID and Auth Token in Settings \u2192 Integrations before deploying.',
        code: 'TWILIO_NOT_CONFIGURED',
      });
    }

    const phoneNumber = await provisionAgentNumber(
      prisma,
      agent_id,
      req.tenantId,
      country_code || 'US',
    );

    // Activate the agent
    await prisma.agent.update({
      where: { id: agent_id },
      data: { status: 'active' },
    });

    res.json({ success: true, phone_number: phoneNumber });
  } catch (error: any) {
    console.error('Error deploying agent:', error);
    res.status(500).json({ error: error?.message || 'Failed to deploy agent' });
  }
});

// Get deployment status
router.get('/status', async (req: Request, res: Response) => {
  try {
    // Return mock status
    res.json({
      status: 'ready',
      message: 'Agent is ready for deployment',
    });
  } catch (error) {
    console.error('Error getting deployment status:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Onboarding progress - simplified version without database
let onboardingProgressStore: { [key: string]: any } = {};

router.post('/progress', async (req: Request, res: Response) => {
  try {
    const { agent_id, current_step, data } = req.body;
    const key = req.userId;

    onboardingProgressStore[key] = {
      agent_id,
      current_step,
      data,
      updatedAt: new Date(),
    };

    res.json({
      success: true,
      agent_id,
      current_step,
      data,
    });
  } catch (error) {
    console.error('Error saving onboarding progress:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.get('/progress', async (req: Request, res: Response) => {
  try {
    const key = req.userId;
    const progress = onboardingProgressStore[key];

    if (progress) {
      res.json({
        exists: true,
        agent_id: progress.agent_id,
        current_step: progress.current_step,
        data: progress.data,
      });
    } else {
      res.json({ exists: false });
    }
  } catch (error) {
    console.error('Error getting onboarding progress:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.delete('/progress', async (req: Request, res: Response) => {
  try {
    const key = req.userId;
    delete onboardingProgressStore[key];
    res.json({ deleted: true });
  } catch (error) {
    console.error('Error deleting onboarding progress:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;

File: new_backend/python-backend/app/routes/onboarding.py
"""
/onboarding routes — mirrors Express src/routes/onboarding.ts
THE most critical routes for the demo flow.
"""
import logging
from typing import Optional, Any
from fastapi import APIRouter, Depends, UploadFile, File, Form, Query, Request
from fastapi.responses import JSONResponse
from sqlalchemy import select, delete as sa_delete
from sqlalchemy.ext.asyncio import AsyncSession
import httpx

from app.database import get_db
from app.auth import AuthContext, get_auth
from app.models import Agent, AgentConfiguration, Tenant, Brand
from app.config import settings

logger = logging.getLogger("voiceflow.onboarding")
router = APIRouter()

# In-memory onboarding progress store (mirrors Express behavior)
_progress_store: dict[str, dict] = {}


# ── Company search (Clearbit) ────────────────────────────────────────────────

@router.get("/company-search")
async def company_search(q: str = Query("")):
    q = q.strip()
    if not q:
        return {"companies": []}
    try:
        async with httpx.AsyncClient(timeout=5) as client:
            resp = await client.get(
                f"https://autocomplete.clearbit.com/v1/companies/suggest?query={q}"
            )
            items = resp.json() if resp.status_code == 200 else []
        companies = [
            {
                "id": item.get("domain") or item.get("name", "").lower().replace(" ", "-"),
                "name": item.get("name", ""),
                "domain": item.get("domain", ""),
                "industry": "",
                "description": item.get("domain", ""),
            }
            for item in (items or [])[:10]
        ]
        return {"companies": companies}
    except Exception:
        logger.exception("Clearbit search failed")
        return {"companies": []}


# ── Company profile ──────────────────────────────────────────────────────────

@router.get("/company")
async def get_company(auth: AuthContext = Depends(get_auth), db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Tenant).where(Tenant.id == auth.tenant_id))
    tenant = result.scalar_one_or_none()
    s = (tenant.settings or {}) if tenant else {}
    return {
        "company_name": s.get("companyName"),
        "industry": s.get("industry"),
        "use_case": s.get("useCase"),
        "website_url": s.get("websiteUrl"),
        "description": s.get("description"),
    }


@router.post("/company")
async def save_company(request: Request, auth: AuthContext = Depends(get_auth), db: AsyncSession = Depends(get_db)):
    body = await request.json()
    company_name = body.get("company_name")
    industry = body.get("industry")
    use_case = body.get("use_case")
    website_url = body.get("website_url")
    description = body.get("description")

    result = await db.execute(select(Tenant).where(Tenant.id == auth.tenant_id))
    tenant = result.scalar_one_or_none()
    if tenant:
        tenant.settings = {
            "companyName": company_name,
            "industry": industry,
            "useCase": use_case,
            "websiteUrl": website_url or None,
            "description": description or None,
        }
        await db.commit()

    scrape_job_id: Optional[str] = None
    if website_url:
        try:
            async with httpx.AsyncClient(timeout=30) as client:
                resp = await client.post(
                    f"{settings.FASTAPI_URL}/ingest/company",
                    json={
                        "tenantId": auth.tenant_id,
                        "website_url": website_url,
                        "company_name": company_name or "",
                        "company_description": description,
                        "industry": industry,
                        "use_case": use_case,
                    },
                )
                scrape_job_id = resp.json().get("job_id")
        except Exception:
            logger.exception("Failed to trigger company scrape")

    return {"success": True, "scrapeJobId": scrape_job_id}


# ── Scrape status proxy ─────────────────────────────────────────────────────

@router.get("/scrape-status/{job_id}")
async def scrape_status(job_id: str):
    try:
        async with httpx.AsyncClient(timeout=10) as client:
            resp = await client.get(f"{settings.FASTAPI_URL}/status/{job_id}")
            return resp.json()
    except Exception:
        return JSONResponse({"error": "Failed to fetch scrape status"}, status_code=500)


# ── Company knowledge proxy ──────────────────────────────────────────────────

@router.get("/company-knowledge")
async def company_knowledge(auth: AuthContext = Depends(get_auth)):
    try:
        async with httpx.AsyncClient(timeout=10) as client:
            resp = await client.get(f"{settings.FASTAPI_URL}/knowledge/company/{auth.tenant_id}")
            return resp.json()
    except Exception:
        return JSONResponse({"error": "Failed to fetch company knowledge"}, status_code=500)


@router.delete("/company-knowledge/{chunk_id}")
async def delete_company_knowledge(chunk_id: str, auth: AuthContext = Depends(get_auth)):
    try:
        async with httpx.AsyncClient(timeout=10) as client:
            await client.delete(f"{settings.FASTAPI_URL}/knowledge/{auth.tenant_id}/{chunk_id}")
        return {"deleted": True}
    except Exception:
        return JSONResponse({"error": "Failed to delete chunk"}, status_code=500)


# ── Agent creation ───────────────────────────────────────────────────────────

@router.post("/agent")
async def create_agent(request: Request, auth: AuthContext = Depends(get_auth), db: AsyncSession = Depends(get_db)):
    body = await request.json()
    name = body.get("name", "Unnamed Agent")
    role = body.get("role")
    template_id = body.get("templateId")
    description = body.get("description")
    channels = body.get("channels")
    brand_id = body.get("brandId")

    if brand_id:
        r = await db.execute(select(Brand).where(Brand.id == brand_id, Brand.tenantId == auth.tenant_id))
        if not r.scalar_one_or_none():
            return JSONResponse({"error": "Brand not found or does not belong to this tenant"}, status_code=400)

    agent = Agent(
        name=name,
        description=description or role or None,
        templateId=template_id or None,
        brandId=brand_id or None,
        channels=channels or None,
        userId=auth.user_id,
        tenantId=auth.tenant_id,
    )
    db.add(agent)
    await db.flush()

    # Get tenant settings for company info
    tr = await db.execute(select(Tenant).where(Tenant.id == auth.tenant_id))
    tenant = tr.scalar_one_or_none()
    tenant_settings = (tenant.settings or {}) if tenant else {}

    config = AgentConfiguration(
        agentId=agent.id,
        templateId=template_id or None,
        agentName=name,
        agentRole=role or None,
        agentDescription=description or None,
        communicationChannels=channels or None,
        companyName=tenant_settings.get("companyName"),
        industry=tenant_settings.get("industry"),
        primaryUseCase=tenant_settings.get("useCase"),
    )
    db.add(config)
    await db.commit()

    return {"agent_id": agent.id}


# ── Knowledge upload ─────────────────────────────────────────────────────────

@router.post("/knowledge")
async def upload_knowledge(
    request: Request,
    auth: AuthContext = Depends(get_auth),
):
    form = await request.form()
    files = form.getlist("files")
    websites = form.get("websites")
    faq_text = form.get("faq_text")

    try:
        async with httpx.AsyncClient(timeout=60) as client:
            data = {"tenant_id": auth.tenant_id, "user_id": auth.user_id}
            if websites:
                data["websites"] = websites
            if faq_text:
                data["faq_text"] = faq_text

            upload_files = []
            for f in files:
                if hasattr(f, "read"):
                    content = await f.read()
                    upload_files.append(("files", (f.filename, content, f.content_type or "application/octet-stream")))

            resp = await client.post(
                f"{settings.FASTAPI_URL}/ingest",
                data=data,
                files=upload_files if upload_files else None,
            )
            job_id = resp.json().get("job_id")
        return {"success": True, "jobId": job_id}
    except Exception:
        logger.exception("Error uploading knowledge")
        return JSONResponse({"error": "Internal server error"}, status_code=500)


# ── Voice configuration ──────────────────────────────────────────────────────

@router.post("/voice")
async def configure_voice(request: Request, auth: AuthContext = Depends(get_auth), db: AsyncSession = Depends(get_db)):
    body = await request.json()
    voice = body.get("voice")
    tone = body.get("tone")
    personality = body.get("personality")

    result = await db.execute(
        select(Agent).where(Agent.tenantId == auth.tenant_id).order_by(Agent.createdAt.desc()).limit(1)
    )
    agent = result.scalar_one_or_none()

    if agent:
        cr = await db.execute(select(AgentConfiguration).where(AgentConfiguration.agentId == agent.id))
        config = cr.scalar_one_or_none()
        if config:
            if voice:
                config.voiceId = voice
            if tone:
                config.responseTone = tone
            if personality:
                config.preferredResponseStyle = personality
        else:
            config = AgentConfiguration(
                agentId=agent.id,
                voiceId=voice or None,
                responseTone=tone or None,
                preferredResponseStyle=personality or None,
            )
            db.add(config)
        await db.commit()

    return {"success": True}


# ── Channel setup ────────────────────────────────────────────────────────────

@router.post("/channels")
async def setup_channels():
    return {"success": True}


# ── Agent configuration ──────────────────────────────────────────────────────

@router.post("/agent-config")
async def agent_config(request: Request, auth: AuthContext = Depends(get_auth), db: AsyncSession = Depends(get_db)):
    body = await request.json()

    result = await db.execute(
        select(Agent).where(Agent.tenantId == auth.tenant_id).order_by(Agent.createdAt.desc()).limit(1)
    )
    agent = result.scalar_one_or_none()

    if agent:
        cr = await db.execute(select(AgentConfiguration).where(AgentConfiguration.agentId == agent.id))
        config = cr.scalar_one_or_none()

        fields = {
            "agentName": body.get("agent_name"),
            "agentRole": body.get("agent_role"),
            "agentDescription": body.get("agent_description"),
            "personalityTraits": body.get("personality_traits"),
            "communicationChannels": body.get("communication_channels"),
            "preferredResponseStyle": body.get("preferred_response_style"),
            "responseTone": body.get("response_tone"),
            "voiceId": body.get("voice_id"),
            "companyName": body.get("company_name"),
            "industry": body.get("industry"),
            "primaryUseCase": body.get("primary_use_case"),
        }

        if config:
            for k, v in fields.items():
                if v is not None:
                    setattr(config, k, v)
        else:
            config = AgentConfiguration(agentId=agent.id, **{k: v for k, v in fields.items() if v is not None})
            db.add(config)
        await db.commit()

    return {
        "success": True,
        "message": "Agent configured successfully",
        "agent_id": agent.id if agent else "unknown",
        "chroma_collection": f"collection_{auth.tenant_id}",
    }


# ── Deploy agent ─────────────────────────────────────────────────────────────

@router.post("/deploy")
async def deploy_agent(request: Request, auth: AuthContext = Depends(get_auth), db: AsyncSession = Depends(get_db)):
    body = await request.json()
    agent_id = body.get("agent_id")
    if not agent_id:
        return JSONResponse({"error": "agent_id is required"}, status_code=400)

    result = await db.execute(select(Agent).where(Agent.id == agent_id, Agent.tenantId == auth.tenant_id))
    agent = result.scalar_one_or_none()
    if not agent:
        return JSONResponse({"error": "Agent not found"}, status_code=404)

    if agent.phoneNumber:
        return {"success": True, "phone_number": agent.phoneNumber, "already_provisioned": True}

    # Check Twilio credentials
    tr = await db.execute(select(Tenant).where(Tenant.id == auth.tenant_id))
    tenant = tr.scalar_one_or_none()
    tenant_settings = (tenant.settings or {}) if tenant else {}
    if not tenant_settings.get("twilioAccountSid") or not tenant_settings.get("twilioAuthToken"):
        return JSONResponse(
            {
                "error": "Twilio credentials not configured. Please add your Twilio Account SID and Auth Token in Settings.",
                "code": "TWILIO_NOT_CONFIGURED",
            },
            status_code=400,
        )

    # In demo mode, just return a mock number
    agent.status = "active"
    await db.commit()
    return {"success": True, "phone_number": "+1-555-DEMO", "already_provisioned": False}


# ── Status ───────────────────────────────────────────────────────────────────

@router.get("/status")
async def get_status():
    return {"status": "ready", "message": "Agent is ready for deployment"}


# ── Onboarding progress (in-memory, same as Express) ────────────────────────

@router.post("/progress")
async def save_progress(request: Request, auth: AuthContext = Depends(get_auth)):
    body = await request.json()
    _progress_store[auth.user_id] = {
        "agent_id": body.get("agent_id"),
        "current_step": body.get("current_step"),
        "data": body.get("data"),
    }
    return {
        "success": True,
        "agent_id": body.get("agent_id"),
        "current_step": body.get("current_step"),
        "data": body.get("data"),
    }


@router.get("/progress")
async def get_progress(auth: AuthContext = Depends(get_auth)):
    progress = _progress_store.get(auth.user_id)
    if progress:
        return {"exists": True, **progress}
    return {"exists": False}


@router.delete("/progress")
async def delete_progress(auth: AuthContext = Depends(get_auth)):
    _progress_store.pop(auth.user_id, None)
    return {"deleted": True}
